doc/metaprogramming.md

```diff
@@ -1 +1,139 @@
 # Metaprogramming
+
+Macros are a language feature that allows writing code that writes other code (metaprogramming).
+
+In Lean 4, macros are used pervasively. So much so that core language features such as `do` notation
+is implemented via macros! As a language user, macros are useful to easily
+embed domain-specific languages and to generate code at compile-time, to name a few uses.
+
+
+#### Balanced Parentheses as an Embedded Domain Specific Language
+
+Let's look at how to use macros to extend the Lean 4 parser and embed a language for building _balanced parentheses_.
+This language accepts strings given by the [BNF grammar](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form)
+
+```
+Dyck := 
+  "(" Dyck ")" 
+  | "{" Dyck '}'
+  | "End"
```

Comment:
This token looks like a hack. How can we make the grammar more conventional? We could use `(brack)?`, but that is not as nice in BNF. An epsilon production would be more common there, but we don't have a direct encoding of that as a `ParserDescr` yet. Should we?

Reply:
Having a `?` would be very useful!

Reply:
We could also modify the `syntax` command to accept an empty sequence of parsers, but an explicit epsilon parser might be the better idea